### ISSUE TITLE
IDEA: User Load Balancing

### ISSUE THREAD

**Issue by djtech-dev** [LemmyNet/lemmy#3825]
### Requirements

- [X] Is this a feature request? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a feature request? Do not put multiple feature requests in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Is your proposal related to a problem?

This proposal is designed to maintain better connections between Instances and allow for distributing the user load.
You can check https://lemmy.world/post/2516669 for the original discussion from where this proposal was written.

### Describe the solution you'd like.

We want to implement an User Load Balancing following the concepts from the discussion linked before.

The implementation should be a public REST api (something like `getUserLoadData`), which takes the address of the server as input [so that you can request the data from the server itself or use a temporary in-memory copy of the data from another server].
The server configurations files will have a `maxUser` variable and `currentUser` will be read in real-time from the server's Database.

From the UI side, I already talked to @jeff-hykin; if this feature is accepted, he'll work on adding the feature in the sign-up section and as a collapsable menu in the sidebar of each instance.

### Describe alternatives you've considered.

After seeing the discussion (both public and in DMs) for this, I think this a good solution for Lemmy's problem of having big instances that can suffer from heavy server workload and a centralized-like environment (like [Element](element.io)).

**Comment by jeff-hykin** [comment]
To follow up with a bit more technical detail, here are the planned changes. (All names are susceptible to change)

Back-end:
- configuration for `targetUserCount`
- configuration for `recommededServers` 
    - list of strings/urls
    - represents who will be recommended once this server hits/exceeds capacity
- configuration for `allowedRecommders`
    - list of strings/urls
    - list of servers that (when THEY are over capacity) are allowed to recommend this server
    - This list is needed to make CORS exception, so that  other servers (e.g. Lemmy.world) can render this servers sign-up page inside of an iframe
- exposing number of users API (`currentUserCount`)
- which server should be recommended (`currentRecommendedServer`)
    - if the current server is under capacity, it should return the URL to the current server
    - otherwise, on a cache-miss/cache-refresh it should query the back-end of each server on its recommeder list, and return the URL to the server that is the most under capacity
    - Note: caching the output for some amount of time; let's say 10min or 10signups, should reduce the number of times servers get pinged and be pretty much irrelevant in terms of behavior.

Front-end:
- The [sign up page](https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/components/home/login.tsx#L23) should be the only one effected
    - If `forceDirectSignup` is a URL parameter, then there is no change to the login page
    - Otherwise the sign up page will ask the back-end for `currentRecommendedServer`
    - If the server recommends itself; then no change to the login page
    - If the server recommends a different server
        - "Sign up using {{other server}}" will be near the top
        - Followed by the other server's sign-up page in an iframe
        - Below the iframe will be "I want to sign up specifically with {{currentServer}}" link, which redirects to the sign-up page with `forceDirectSignup` URL parameter

**Comment by Nutomic** [comment]
Sounds extremely complex and I have no idea what problem it would solve.

**Comment by jeff-hykin** [comment]
:/ Why close this without discussion?



![Screenshot_20230929-074147](https://github.com/LemmyNet/lemmy/assets/17692058/1983e5f2-035f-4f3f-bba7-24f49e747714)

Lemmy.world might be able to handle hundreds of thousands of users, but that doesnt mean all servers have unlimited capacity. My friends already find signing up for lemmy on a non-main instance annoying enough to just not get an acocunt. This change solves two sides of that problem; a better UX for users trying to find a smaller instance, and a better way for servers to stay at/under capacity. 

If you want to say "the technical overhead is too much for this feature" I understand. But saying "TLDR; closing" feels pretty bad.

The changes are not complex, I just put a lot of detail.

TLDR;
- two config options (desiredNumberOfUsers, recommendedServers)
- 1 endpoint for each of ^
- if over capacity, render another server's login at the top

Thats it.

**Comment by Nutomic** [comment]
Okay thats much easier to understand. But do you have any indication that lemmy.world would use such a feature? They could simply close registrations if they were overloaded, but havent done that. And it might be enough to simply put a link to the joinlemmy instance list on the registration page if signups are closed.

If people want to signup despite closed registrations, that could be handled by implementing invite links. These things seem much easier and more effective to implement.

**Comment by jeff-hykin** [comment]
> But do you have any indication that lemmy.world would use such a feature?

Sadly I do not :/ and from what little I've seen, I would get they'd consider themselves under capacity. 


> it might be enough to simply put a link to the joinlemmy instance list on the registration page if signups are closed

I mean, that sounds like a good idea regardless of this feature! But still every click looses like 65% of people, so adding three extra clicks (closed-registration->join lemmy page->lemmy with open registration) is going to loose a lot. 

I'd like to host my own lemmy server, post the link to my signup page in different locations, and know that the server just automatically render my recommended servers as soon as I hit my user limit.
